Integrate visibles analysis tools into ldmx-sw

horohoo opened the pull request:
I am updating _ldmx-sw_, here are the details.

### What are the issues that this addresses?
<!--
_Hint_: Use the phrase '_This resolves #< issue number >_' so that they are linked automatically.
-->
This resolves #1478 which requests that the visibles analysis be integrated into _ldmx-sw_.

## Check List
- [x] I successfully compiled _ldmx-sw_ with my developments.
- [x] I read, understood and follow the [coding rules](https://ldmx-software.github.io/developing/coding-rules.html).
<!-- If these coding rules are confusing or you need help, still open the PR as a _draft_ and we can help you! -->
- [x] I ran my developments and the following shows that they are successful.
<!-- put plots or some other proof that your developments work and do the intended function -->
Not sure of the best plot to put here to show that the developments work. There are several that I could use, but I think the best demonstrator is that the visibles BDT in this framework (with other selection cuts) reproduces the efficiency we expect from previous studies. A more thorough set of plots are included in the draft of the visibles paper, which I can link to a Docdb entry soon.
<img width="966" height="730" alt="efficiency" src="https://github.com/user-attachments/assets/91fc7e59-8789-4455-8ae7-83c78b50fddb" />

horohoo committed:
First steps to build a visibles veto processor

horohoo committed:
Committing all changes to VisiblesVetoProcessor

horohoo committed:
Updated configuration for VisiblesVetoProcessor

horohoo committed:
Initial commit of Visibles training script

horohoo committed:
Adding capability to train BDT in VisiblesVetoProcessor

horohoo committed:
Added framework for handling rec vs truth tracks

horohoo committed:
Some minor changes to the workflow

horohoo committed:
Basic version of the BDT training script

horohoo committed:
Small patches that finally get the code to compile

horohoo committed:
Adding Event/VisiblesVetoResult to CMakeLists

horohoo committed:
Moved the job of training to another class

Adding the BDT model

horohoo committed:
Moving visibles BDT training to its own class

horohoo committed:
BDT threshold set

horohoo committed:
Include recoil tracking cut

horohoo committed:
A class to run visibles analysis

horohoo committed:
Slightly improved BDT performance

horohoo committed:
Add pass name to all collections

horohoo committed:
Add EaT option for testing

horohoo committed:
Require high momentum recoil electron for EaT

horohoo committed:
Removing EaT stuff, should be in new classes

horohoo committed:
Adding FPR/TPR capability to cutflow

horohoo committed:
Oops, forgot to remove an EaT condition

horohoo committed:
Swap abs for fabs

horohoo committed:
Cleaning up script

horohoo committed:
Update model in VisiblesVetoProcessor

horohoo committed:
Set an arbitrary value to gamma_E

horohoo committed:
Fixing all abs to fabs issues

github-actions[bot] committed:
Apply clang-format

Removing redundant visibles BDT model

tvami commented on Hcal/data/visibles_v1.onnx:
This file is not used, right? We should not add it, and even more it should be removed from the git history too (by doing a squash of the commit that added it and the one that removed it) -- Given it's size is non-negligible

horohoo replied:
I think this is done now. I have not done a squash before, so whether this was done properly is subject to the accuracy of ChatGPT's instructions.

tomeichlersmith replied:
GitHub's squash merge would remove it from the history when you click the "Squash and Merge" button after review. It looks like ChatGPT's instructions were not very accurate since you now have 499 files you are changing.

horohoo replied:
Okay, let me undo ChatGPT's mess and we can go from there. There are two separate commits that add the redundant file and remove it, so that makes things simple when it's time to squash.

tvami commented on Hcal/include/Hcal/Event/VisiblesVetoResult.h:
```diff
@@ -0,0 +1,85 @@
+  /** Destructor */
+  virtual ~VisiblesVetoResult();
+
+  void setVariables(int nLayersHit, double xStd, double yStd, double zStd,
```
Please change the variable names to it matches the coding rules: https://ldmx-software.github.io/developing/coding-rules.html#i-naming-conventions

tvami commented on Hcal/include/Hcal/Event/VisiblesVetoResult.h:
```diff
@@ -0,0 +1,85 @@
+
+  void Print() const;
+
+  bool passesVeto() const { return passesVeto_; }
```
```suggestion
  bool passesVeto() const { return passes_veto_; }
```
etc, same comment about naming conventions

tvami replied:
This is a comment for the whole PR, but I wont point it out for all the cases, please just apply the rules everywhere

tvami commented on Hcal/include/Hcal/Event/VisiblesVetoResult.h:
```diff
@@ -0,0 +1,85 @@
+
+  double discValue_{0};
+
+  ClassDef(VisiblesVetoResult, 7);
```
```suggestion
  ClassDef(VisiblesVetoResult, 1);
```

tvami commented on Hcal/include/Hcal/VisiblesCutflow.h:
```diff
@@ -0,0 +1,48 @@
+
+ private:
+  std::unique_ptr<ldmx::Ort::ONNXRuntime> rt_;
+  double bdtCutVal_{0.};
```
```suggestion
  double bdt_cut_val_{0.};
```
again naming in this file too

tvami commented on Hcal/src/Hcal/VisiblesFeatureProducer.cxx:
```diff
@@ -0,0 +1,330 @@
+                                                   sp_pass_name_);
+      bool foundRec = false;
+      for (auto const &it : particle_map) {
+        std::cout << "PDG ID: " << it.second.getPdgID()
+                  << " , Energy: " << it.second.getEnergy() << std::endl;
```
```suggestion
        lmdx_log(trace) << "PDG ID: " << it.second.getPdgID()
                  << " , Energy: " << it.second.getEnergy();
```

tvami commented on Hcal/src/Hcal/VisiblesFeatureProducer.cxx:
```diff
@@ -0,0 +1,330 @@
+    if (training_) {
+      std::ofstream file(trainingFile_, std::ios::app);
+      if (!file.is_open()) {
+        std::cerr << "Error: Could not open file " << trainingFile_
+                  << std::endl;
```
This should be in the logging system too, prob `fatal`

tvami commented on Hcal/src/Hcal/VisiblesCutflow.cxx:
```diff
@@ -0,0 +1,407 @@
+
+void VisiblesCutflow::configure(framework::config::Parameters &parameters) {
+  rt_ = std::make_unique<ldmx::Ort::ONNXRuntime>(
+      parameters.getParameter<std::string>("bdt_file"));
```
```suggestion
      parameters.get<std::string>("bdt_file"));
```
You can do the same for all the `getParameter`-s, it's not a requirement but we are transitioning to this simpler version

tvami reviewed (changes requested):
Thanks Tyler! I read half of it, but it's already enough for sending it back as a first pass. It's mostly about our naming, the log usage, and adding unused files.

horohoo commented:
>was the BDT trained with scoring plane information

Yes, our training samples did not have reconstructed tracker hits, so we trained on target scoring plane hits.

I will take a look at the CIs and see what makes sense for testing on this processor. When I make changes to it, I'll commit separately from the other requested changes.

horohoo committed:
Rename visibles BDT model

horohoo committed:
Reformat variable names, update BDT path in visibles.py

github-actions[bot] committed:
Apply clang-format

horohoo committed:
Catching all missed variable name changes

horohoo commented:
It looks like the ONNX we want removed actually did get removed. Previously, there were two models: `visibles_v1.onnx` and `visibles_v2.onnx`. In two separate commits, I removed `visibles_v1.onnx` (the one we don't need) and renamed `visibles_v2.onnx` to `visibles.onnx`. So all of the differences between `iss1478` and `trunk` are the ones we want.
But since I'm seeing
```
-bash-4.4$git status
On branch iss1478
Your branch and 'origin/iss1478' have diverged,
and have 84 and 86 different commits each, respectively.
  (use "git pull" if you want to integrate the remote branch with yours)

nothing to commit, working tree clean
```
won't a push still include all of the commits to stuff that are other already-`trunk`-merged commits?

tomeichlersmith commented:
:tada: Hooray :tada:

<img width="651" height="67" alt="image" src="https://github.com/user-attachments/assets/8c5c8a14-8d12-400b-b48c-ad674e32d8b6" />

:muscle: Humans Helping Humans :muscle:

tvami commented on Hcal/python/TrainVisiblesBDT.py:
```diff
@@ -0,0 +1,321 @@
+              "early_stopping_rounds": 10}
+
+    evallist = [(eventContainer.dtest, 'eval'), (eventContainer.dtrain, 'train')]
+    gbm = xbg.train(params, eventContainer.dtrain, options.tree_number, evallist)
```
```suggestion
    gbm = xgb.train(params, eventContainer.dtrain, options.tree_number, evallist)
```
Did the original really work?

horohoo replied:
I have been neglecting this file. I will overhaul it with our group's up-to-date training method, which doesn't rely on looping over events in Python! Does a script for training the visibles BDT belong in the Hcal package or is there a better place for it? It could go outside of `ldmx-sw` and instead go in `ldmx-publications` as an analysis script, like [here](https://github.com/LDMX-Software/ldmx-publications/tree/2025visibles/2025-visibles/analysis-code) (more will change in that repo to reflect how this PR is integrated into `ldmx-sw`).

tomeichlersmith replied:
I think it makes more sense for it to be outside ldmx-sw since it is not related to the C++ processing (but we also have other stuff that's not related to the C++ processing floating around).

tvami commented on Hcal/include/Hcal/VisiblesVetoProcessor.h:
```diff
@@ -0,0 +1,80 @@
+
+  double beam_energy_mev_{0};
+
+  bool verbose_{false};
```
```suggestion
```

tvami commented on Hcal/src/Hcal/VisiblesVetoProcessor.cxx:
```diff
@@ -0,0 +1,256 @@
+
+void VisiblesVetoProcessor::configure(
+    framework::config::Parameters &parameters) {
+  verbose_ = parameters.get<bool>("verbose");
```
```suggestion
```

tvami commented on Hcal/src/Hcal/VisiblesVetoProcessor.cxx:
```diff
@@ -0,0 +1,256 @@
+      n_readout_hits_ += 1;
+      double x = hit.getXPos();
+      double y = hit.getYPos();
+      double z = hit.getZPos();
```
Please avoid one character variable names (including the `r`, see I.1 in https://ldmx-software.github.io/developing/coding-rules.html

tvami commented on Hcal/src/Hcal/VisiblesVetoProcessor.cxx:
```diff
@@ -0,0 +1,256 @@
+      double x = hit.getXPos();
+      double y = hit.getYPos();
+      double z = hit.getZPos();
+      double r = sqrt(pow(x, 2) + pow(y, 2));
```
```suggestion
      double r = sqrt(x*x + y*y);
```
is much faster (but also please rename x [maybe `pos_x`])

tvami commented on Hcal/src/Hcal/VisiblesVetoProcessor.cxx:
```diff
@@ -0,0 +1,256 @@
+      double y_proj = gamma_x0[1] + (z - gamma_x0[2]) * gamma_p[1] / gamma_p[2];
+
+      r_mean_from_photon_track_ +=
+          hit.getEnergy() * sqrt(pow(x - x_proj, 2) + pow(y - y_proj, 2));
```
same comment about not using pow()

tvami commented on Hcal/src/Hcal/VisiblesVetoProcessor.cxx:
```diff
@@ -0,0 +1,256 @@
+            // Determine if a bar is vertical (along y-axis) or horizontal
+            // (along x-axis) Odd layers have horizontal strips Even layers have
+            // vertical strips
+            if (detID2.getLayerID() % 2 == 0) {
```
I dont know if this ID check will be always true or not? 
There are methods on the hit itself to know it's orientation:
https://github.com/LDMX-Software/ldmx-sw/blob/28fdf455055fd6b69e958b8779ae8c8248a564df/Hcal/include/Hcal/Event/HcalHit.h#L121-L137

that may simplify this code?

tvami commented on Hcal/python/visibles.py:
```diff
@@ -0,0 +1,159 @@
+        self.sp_pass_name = ''
+        self.sim_particles_pass_name = ''
+
+class VisiblesFeatureProducer(ldmxcfg.Analyzer) :
```
Should this be under the DQM package?

horohoo replied:
I think if any class belongs in DQM, it is `VisiblesCutflow`, which reproduces all of the analysis cuts for the visibles analysis and produces relevant histograms. I agree the `VisiblesFeatureProducer` class seems out of place in the Hcal package. Its function is to produce BDT feature histograms or files for training a new BDT. Would DQM be an appropriate place for this? Could Tools also be appropriate?

tvami replied:
Yeah I think we have the all the Ecal + Tracker (and some of the HCAL too) stuff that does similar things like the `VisiblesCutflow` and `VisiblesFeatureProducer` under DQM

horohoo replied:
Okay, thanks. I will move both classes into DQM.

tvami replied:
Thanks!

tvami commented on Hcal/python/visibles.py:
```diff
@@ -0,0 +1,159 @@
+        ## Define histograms ##
+
+        self.build1DHistogram("layers_hit", "layers_hit", 100, 0, 100);
+        self.build1DHistogram("x_std", "x_std", 80, 0, 800);
```
Full the full next block, I'd like to ask you to include nicer axis titles, especially to include units, in this specific case that the x std is in mm, I assume?

tvami commented on Hcal/src/Hcal/VisiblesFeatureProducer.cxx:
```diff
@@ -0,0 +1,323 @@
+          continue;
+        }
+
+        n_readout_hits += 1;
+        double x = hit.getXPos();
+        double y = hit.getYPos();
+        double z = hit.getZPos();
+        double r = sqrt(pow(x, 2) + pow(y, 2));
```
Same comments here in the VetoProcessor

tvami reviewed (changes requested):
Great, thanks Tyler for the prev changes. Here is my last round review

horohoo committed:
Moving visibles classes into DQM

horohoo committed:
Changed hit variable names, swapped out pow function

horohoo committed:
Update determination of Hcal bar orientation

horohoo committed:
Remove verbose

horohoo committed:
Removed visibles BDT training script for a new repo home

github-actions[bot] committed:
Apply clang-format

tvami commented:
OK, so the only thing missing is to add this to one of the CIs, right?

horohoo committed:
Add option to turn off other cuts and only check visibles BDT

horohoo committed:
Add test of visibles BDT to ecal PN validation sample

github-actions[bot] committed:
Apply clang-format

horohoo commented:
It wouldn't hurt to add `VisiblesFeatureProducer` as well. `VisiblesCutflow` also produces a set of feature plots, which is why I initially didn't include `VisiblesFeatureProducer`, but they will be slightly different due to the lack of event selection in `VisiblesCutflow`.

horohoo committed:
Add VisiblesFeatureProducer to validation sample

tvami reviewed (approved):
Looks really good, thanks for the nice work Tyler! I just initiated the test with the CI that includes the changes

tomeichlersmith commented on DQM/src/DQM/VisiblesCutflow.cxx:
```diff
@@ -0,0 +1,410 @@
+      in_list = true;
+    }
+  }
+  return in_list;
```
Would need to `#include <algorithm>`
```suggestion
return std::find(parents.begin(), parents.end(), track_id) != parents.end();
```

tomeichlersmith commented on DQM/src/DQM/VisiblesCutflow.cxx:
```diff
@@ -0,0 +1,410 @@
+                   std::acos(gamma_p[2] / std::sqrt(gamma_p[0] * gamma_p[0] +
+                                                    gamma_p[1] * gamma_p[1] +
+                                                    gamma_p[2] * gamma_p[2])));
+  ;
```
```suggestion
```

tomeichlersmith commented on DQM/src/DQM/VisiblesCutflow.cxx:
```diff
@@ -0,0 +1,410 @@
+
+  double decay_z;
+
+  auto particle_map{event.getMap<int, ldmx::SimParticle>(
```
Making this a `const` reference `&` prevents an extra copy on each event. If you want to edit the map/vector while filtering it, then making a copy makes sense. Just wanting to avoid it if its not necessary.

```suggestion
  const auto &particle_map{event.getMap<int, ldmx::SimParticle>(
```

tomeichlersmith commented on DQM/src/DQM/VisiblesCutflow.cxx:
```diff
@@ -0,0 +1,410 @@
+  bool found_recoil_e = false;
+
+  if (recoil_from_tracking_) {
+    auto recoil_tracks{
```
Same comment as the particle map
```suggestion
    const auto &recoil_tracks{
```

tomeichlersmith commented on DQM/src/DQM/VisiblesCutflow.cxx:
```diff
@@ -0,0 +1,410 @@
+    }
+  } else {
+    if (event.exists(sp_collection_, sp_pass_name_)) {
+      std::vector<ldmx::SimTrackerHit> target_sp_hits =
```
```suggestion
      const auto &target_sp_hits =
```

tomeichlersmith commented on DQM/src/DQM/VisiblesCutflow.cxx:
```diff
@@ -0,0 +1,410 @@
+  histograms_.fill("pass_acceptance", decay_z);
+
+  // Get EcalRecHits, check that trigger is passed
+  std::vector<ldmx::EcalHit> ecal_rec_hits = event.getCollection<ldmx::EcalHit>(
```
```suggestion
  const auto &ecal_rec_hits = event.getCollection<ldmx::EcalHit>(
```

tomeichlersmith commented on DQM/src/DQM/VisiblesCutflow.cxx:
```diff
@@ -0,0 +1,410 @@
+  // Get EcalRecHits, check that trigger is passed
+  std::vector<ldmx::EcalHit> ecal_rec_hits = event.getCollection<ldmx::EcalHit>(
+      ecal_rec_collection_, ecal_rec_pass_name_);
+  std::vector<ldmx::HcalHit> hcal_rec_hits = event.getCollection<ldmx::HcalHit>(
```
```suggestion
  const auto &hcal_rec_hits = event.getCollection<ldmx::HcalHit>(
```

tomeichlersmith commented on DQM/src/DQM/VisiblesFeatureProducer.cxx:
```diff
@@ -0,0 +1,328 @@
+void VisiblesFeatureProducer::analyze(const framework::Event &event) {
+  std::vector<double> bdt_features;
+
+  auto particle_map{event.getMap<int, ldmx::SimParticle>(
```
Same comment about getting the maps/collections as the other file (I'm just too lazy to put in suggestions).

tomeichlersmith commented on DQM/src/DQM/VisiblesFeatureProducer.cxx:
```diff
@@ -0,0 +1,328 @@
+    bdt_features.push_back(r_mean_from_photon_track);
+
+    if (training_) {
+      std::ofstream file(training_file_, std::ios::app);
```
Why do you have a separate text CSV file for training?

I feel like we could put the bdt features into the event bus and then use `uproot` on the Python side to load these arrays into a table (probably pandas.Dataframe or pytorch.Tensor). This comment is not blocking this PR, I'm just using it as an opportunity to point out that you can add "simple" objects to the event bus and use them later.

horohoo replied:
The honest answer for why we used separate text CSVs for training is that I don't know all the features of the event bus :)
This method of saving/loading CSVs could definitely be deprecated in a future iteration of the visibles analysis workflow.

tomeichlersmith replied:
No worries, sounds like a good plan :)

horohoo committed:
Streamline inList function

horohoo committed:
Load collections as const references

github-actions[bot] committed:
Apply clang-format

tvami commented:
Tests pass nicely, the failure in target_pn_lyso is not related to this PR
